Add token templating

zarvox opened the pull request:
Grains can now make a request to the shell via postMessage() with a template
which can substitute a privileged API token for $API_TOKEN and the API host for
$API_HOST.  The grain will receive a postMessage reply from the shell with a
single-use URL in the shell's origin that will display the rendered template.
The app should iframe this URL in a place in the document inline where it wants
the template to appear.  Apps should *not* persist this URL anywhere.

Implementation detail: the privileged token will expire 5 minutes after
creation, if it is not used in that timeframe.  This helps avoid cluttering the
DB (and UI) with a bunch of tokens that were generated for this templating magic
that the user didn't end up using.

zarvox commented:
Failed due to [a bug in meteor](https://github.com/meteor/meteor/issues/3877).  Workaround is in testing.

paulproteus commented:
Good spotting.​

paulproteus commented:
Garply, retest this please

kentonv commented on shell/public/tokentemplate.html:
```diff
@@ -0,0 +1,48 @@
+<script>
+// Templates are namespaced in localStorage by the prefix "template"
+var templateToken = "template" + window.location.hash.substring(1);
+var record = JSON.parse(localStorage.getItem(templateToken));
```
Use `sessionStorage` rather than `localStorage`? Then you don't really need to clean up expired templates.

kentonv commented on shell/public/tokentemplate.html:
```diff
@@ -0,0 +1,48 @@
+    <pre id="text"></pre>
+<script>
+// Templates are namespaced in localStorage by the prefix "template"
+var templateToken = "template" + window.location.hash.substring(1);
```
How about "offerTemplate", to be somewhat more specific? Filename could be `offer-template.html`.

kentonv commented on shell/server/hack-session.js:
```diff
@@ -270,12 +276,22 @@ Meteor.methods({
+  ApiTokens.remove({expiresIfUnused: {$lt: now}});
+};
+
+// Make self-destructing tokens actually self-destruct, so they don't
```
Could we also not display the token in the webkey UI until it has had its first use?

kentonv commented on shell/shared/grain.js:
```diff
@@ -511,6 +511,60 @@ if (Meteor.isClient) {
         Session.set("grainFrameTitle", event.data.setTitle);
+      } else if (event.data.renderTemplate) {
+        // Request creation of a single-use template with a privileged API token.
+        // Why?  Apps should not be able to obtain capabilities to themselves
```
Specifically, capabilities-as-keys to themselves, that is capabilities that can be leaked through an arbitrary bit stream.

kentonv commented on shell/shared/grain.js:
```diff
@@ -511,6 +511,60 @@ if (Meteor.isClient) {
+          check(call.petname, String);
+          petname = call.petname;
+        } else {
+          petname = "template token";
```
Users aren't likely to know what this means. Is there anything more understandable we could use? Maybe just "connected external app"?

kentonv commented on shell/public/tokentemplate.html:
```diff
@@ -0,0 +1,48 @@
+// Templates are namespaced in localStorage by the prefix "template"
+var templateToken = "template" + window.location.hash.substring(1);
+var record = JSON.parse(localStorage.getItem(templateToken));
+console.log(record);
```
Remove debug logging?

kentonv commented on shell/shared/grain.js:
```diff
@@ -511,6 +511,60 @@ if (Meteor.isClient) {
+        } else {
+          petname = "template token";
+        }
+        // Tokens expire by default in 5 minutes from generation date
```
Let's make sure to implement some sort of periodic ping from the template HTML that prevents the token from expiring while it's still open.

zarvox replied:
Hmm, is that possible without making the template load all of Meteor's JS to get an authenticated connection?  At that point, does it make more sense to make the currently-static-HTML page become just another route on the meteor app?

zarvox commented:
Addressed all comments, except for the "unused offer token iframe sends keepalives while still visible" request, which I'm not sure offhand how to actually integrate in both a lightweight and safe fashion.  At present, the template is short and entirely self-contained; sending keepalives would require Meteor and a session and so on.

It should be possible to add the keepalives at a later date without breaking any of the current APIs or behaviors, though, so perhaps we could punt on that and revisit later?

kentonv commented:
I'm fine with not adding keepalives in this PR. But I think it will be easy. You don't need a Meteor context. Instead, we just need to expose a way to send an XHR to the API endpoint which refreshes the timeout to 5 minutes. There's no security concerns, since a similar XHR could just as easily count as the token's first use, keeping it alive permanently.

dwrensha commented on shell/shared/grain.js:
```diff
@@ -511,6 +511,62 @@ if (Meteor.isClient) {
+                                           .replace("$API_HOST", makeWildcardHost("api"));
+            sessionStorage.setItem(key, JSON.stringify({
+                "renderedTemplate": renderedTemplate,
+                "expiry": selfDestructTime
```
It looks like the code in offer-template.html expects an "expires" field, not an "expiry" field.

zarvox replied:
Whoops, you're right.  Good thing this is strictly advisory for now!

dwrensha commented on shell/shared/grain.js:
```diff
@@ -511,6 +511,62 @@ if (Meteor.isClient) {
+            var tokenId = result.token;
+            // Generate random key id2.
+            var id2 = Random.secret();
+            // Store apitoken id1 and template in local storage in the offer
```
s/local storage/session storage/

zarvox replied:
another good catch

dwrensha commented on shell/shared/grain.js:
```diff
@@ -511,6 +511,62 @@ if (Meteor.isClient) {
+        var template = call.template;
+        check(template, String);
+        var assignment = undefined;
+        if (call.roleId) {
```
Does this mean that these templates can't be used for `allAccess` tokens? Is there any reason to allow an arbitrary Grain.ViewSharingLink.RoleAssignment here?

zarvox replied:
I'm not sure I know the meaning of allAccess tokens yet - is that shorthand for "all permissions defined in this app's pkgdef.capnp"?

As for having a particular role rather than allAccess: in the case of Piwik, we'll want to use the token to allow the pages with tracking code to submit pageview events, but probably don't want to grant that token access to view config data, or actual site statistics.  So I think supporting some limits to the permissions granted is necessary.

dwrensha replied:
Oops, sorry. I meant to ask "Is there any reason **not** to allow an arbitrary Grain.ViewSharingLink.RoleAssignment here?"

The allAccess role carries all permissions. This means that whoever uses the token will get exactly the permissions of the user who created the token. As discussed in https://github.com/sandstorm-io/sandstorm/pull/443, this is commonly the right thing.
